From 0bc95c48c7ce7d3be741c99b464f44ea25be834e Mon Sep 17 00:00:00 2001
From: Edward Wu <edwardwu@realtek.com>
Date: Wed, 14 Aug 2024 07:02:18 +0800
Subject: [PATCH 06/10] linux-5.15: rtk_pm: handle pcpu_param->wakeup_source
 with mask

- wakeup: LAN, IRDA, CEC
-- LAN_EVENT, IR_EVENT, CEC_EVENT
-- CONFIG_R8169SOC, CONFIG_IR_RTK, CONFIG_RTK_HDMI_CEC

Change-Id: Ic09d9830d017ee15a25aa9a05ded4cb15e6b1e52
---
 drivers/soc/realtek/common/rtk_pm.c        |  2 ++
 drivers/soc/realtek/common/rtk_pm_common.c | 15 +++++++++++++++
 drivers/soc/realtek/common/rtk_pm_sysfs.c  |  5 +++++
 include/soc/realtek/rtk_pm.h               |  1 +
 4 files changed, 23 insertions(+)

Index: kernel-5_15/drivers/soc/realtek/common/rtk_pm.c
===================================================================
--- kernel-5_15.orig/drivers/soc/realtek/common/rtk_pm.c
+++ kernel-5_15/drivers/soc/realtek/common/rtk_pm.c
@@ -81,6 +81,8 @@ static int rtk_pm_prepare(struct device
 	struct pm_dev_param *lan_node = rtk_pm_get_param(LAN);
 	int ret = 0;
 
+	dev_pm->pcpu_param->wakeup_source &= rtk_pm_get_param_mask();
+
 	if (lan_node == NULL)
 		goto skip_set_dco_param;
 
Index: kernel-5_15/drivers/soc/realtek/common/rtk_pm_common.c
===================================================================
--- kernel-5_15.orig/drivers/soc/realtek/common/rtk_pm_common.c
+++ kernel-5_15/drivers/soc/realtek/common/rtk_pm_common.c
@@ -40,6 +40,21 @@ struct pm_dev_param *rtk_pm_get_param(un
 }
 EXPORT_SYMBOL(rtk_pm_get_param);
 
+unsigned int rtk_pm_get_param_mask(void)
+{
+	/* CONFIG_R8169SOC, CONFIG_IR_RTK, CONFIG_RTK_HDMI_CEC */
+	unsigned long param_mask = ~(BIT(LAN_EVENT) | BIT(IR_EVENT) | BIT(CEC_EVENT));
+	struct pm_dev_param *node;
+
+	list_for_each_entry(node, &rtk_pm_param_list, list) {
+		if (node->dev_type)
+			param_mask |= BIT(node->dev_type - 1);
+	}
+
+	return (unsigned int)param_mask;
+}
+EXPORT_SYMBOL(rtk_pm_get_param_mask);
+
 void rtk_pm_set_pcpu_param(struct device *dev)
 {
 	struct pm_private *dev_pm = dev_get_drvdata(dev);
Index: kernel-5_15/drivers/soc/realtek/common/rtk_pm_sysfs.c
===================================================================
--- kernel-5_15.orig/drivers/soc/realtek/common/rtk_pm_sysfs.c
+++ kernel-5_15/drivers/soc/realtek/common/rtk_pm_sysfs.c
@@ -156,8 +156,12 @@ static ssize_t rtk_pm_wakeup_show(struct
 	struct pm_dev_param *pm_param = rtk_pm_get_param(PM);
 	struct pm_private *dev_pm = dev_get_drvdata(pm_param->dev);
 	unsigned int n = 0;
+	unsigned int param_mask = rtk_pm_get_param_mask();
 
 	for (i = 0; i < MAX_EVENT; i++) {
+		if (!(param_mask & BIT(i)))
+			continue;
+
 		if (dev_pm->pcpu_param->wakeup_source & BIT(i))
 			n += snprintf(buf + n, PAGE_SIZE, " * ");
 		else
@@ -181,6 +185,7 @@ static ssize_t rtk_pm_wakeup_store(struc
 
 	if (wakeup < MAX_EVENT) {
 		dev_pm->pcpu_param->wakeup_source ^= BIT(wakeup);
+		dev_pm->pcpu_param->wakeup_source &= rtk_pm_get_param_mask();
 		return count;
 	}
 
Index: kernel-5_15/include/soc/realtek/rtk_pm.h
===================================================================
--- kernel-5_15.orig/include/soc/realtek/rtk_pm.h
+++ kernel-5_15/include/soc/realtek/rtk_pm.h
@@ -58,6 +58,7 @@ struct mem_check {
 extern void rtk_pm_init_list(void);
 extern void rtk_pm_add_list(struct pm_dev_param *pm_node);
 extern struct pm_dev_param *rtk_pm_get_param(unsigned int id);
+extern unsigned int rtk_pm_get_param_mask(void);
 extern int rtk_pm_create_sysfs(void);
 extern void rtk_pm_set_pcpu_param(struct device *dev);
 extern void rtk_pm_del_list(struct pm_dev_param *pm_node);
